### ISSUE TITLE
crash upon high concurrent connections (when using the select backend)

### ISSUE THREAD

**Issue by devlearner** [h2o/h2o#287]
Hi kazuho

Sorry for a new issue again :)

I recently tested with h2load against h2o, h2o seems to crash upon high concurrent clients (>54).
The h2o webserver is serving local files at 127.0.0.1:555

I tested h2o with h2load (without tls):
```
$ h2load -c 54 -n 54 -v http://127.0.0.1:555/text.txt
```
(`text.txt` is just a test text file with the content `abc`)

The 'safe-zone' seems to be when concurrent clients (`-c`) are between 1 - 54. 

When concurrent clients = 55
```
$ h2load -c 55 -n 55 -v http://127.0.0.1:555/text.txt
```

h2o crashes with the following message:
```
assertion "sock->super._cb.write != NULL" failed: file " (...) /h2o-1.2.0/lib/common/socket/evloop.c.h", 
line 179, function: write_pending
```

When concurrent clients >= 56
```
$ h2load -c 56 -n 56 -v http://127.0.0.1:555/text.txt
...
assertion "loop->socks[sock->fd]->_flags == H2O_SOCKET_FLAG_IS_DISPOSED" failed: file " (...) 
/h2o-1.2.0/lib/common/socket/evloop/select.c.h", line 143, function: evloop_do_on_socket_create
```

Sometimes it shows this message:
```
Program received signal SIGSEGV, Segmentation fault.

evloop_do_on_socket_create (sock=<optimized out>, sock=<optimized out>)
    at (...) /h2o-1.2.0/lib/common/socket/evloop/select.c.h:143
143             assert(loop->socks[sock->fd]->_flags == H2O_SOCKET_FLAG_IS_DISPOSED);
```

which seems to point to this secton in `/lib/common/socket/evloop/select.c.h`:
```
(gdb) list
138
139         if (loop->max_fd < sock->fd)
140             loop->max_fd = sock->fd;
141
142         if (loop->socks[sock->fd] != NULL) {
143             assert(loop->socks[sock->fd]->_flags == H2O_SOCKET_FLAG_IS_DISPOSED);
144         }
145         assert(!FD_ISSET(sock->fd, &loop->readfds));
146         assert(!FD_ISSET(sock->fd, &loop->writefds));
147     }
```

[I'm on a platform known not to handle high concurrent connections very well ;) ]
but since it seems to be quite reproducible (interestingly at concurrent clients >=55) in my case, and just for reference, I tested another http2 implementation which seems to be handling concurrent clients = 1000 quite well, I suspect that it might not be running out of sockets/ file descriptors in this case.

of course it doesn't pose a huge inconvenience to me in my use case (doesn't really get to that load)
it would be very kind if you might take a look at it (of course, at a time convenient to you ;))

thanks a lot!

**Comment by kazuho** [comment]
@devlearner What OS are you using?

H2O selects the `epoll` backend or `kqueue` backend is selected for most operating systems, unless specified otherwise (see https://github.com/h2o/h2o/blob/master/lib/common/socket/evloop.c.h#L59).

However, the fact that `lib/common/socket/evloop/select.c.h` is being used indicates that the _select_ backend for the embedded event loop is being used.  The select backend exists for mainly test and compatibility purposes, and is known to have an issues in harding more than certain number of connections (i.e. maximum fd should not become greater than `FD_MAXSIZE`).

**Comment by devlearner** [comment]
Thanks for the reply (am using cygwin)
Seems like maybe it's an inherent limitation of the `select` backend after-all.
Thanks for the help ;)

**Comment by kazuho** [comment]
@devlearner Had not thought about Cygwin.  Created a new issue that would fix the problem once implemented.

**Comment by devlearner** [comment]
Found a partial workaround!

Thanks kazuho for pointing me to the maximum fd thing. Turns out it's partially platform-specific:

Found out that (after some googling) cygwin has a default FD_SETSIZE as low as 64
probably due to some historic reasons (which is... um, pathetically low, IMO)

Recompiled h2o with a raised FD_SETSIZE flag: `-DFD_SETSIZE=1024` or like `-DFD_SETSIZE=65536`
and now h2o is taking 1000 concurrent connections without the seg fault ;)
... which is quite impressive indeed! considering the use of `select` backend on my platform 

(Mileage will vary; but it peaks at concurrent connections = 1224 on my platform, just for the record)

[I should have figured this out early; probably the other http2 implementation I tested uses libev and is raising this at code level]

That said, thumbs-up for the `poll` backend if h2o is going to get one; that'll probably help other platforms like Solaris!

Big thanks for all the help !
